Dynamically expose attributes

vk496 opened the pull request:
Hello,

As already discussed in https://github.com/Koenkk/zigbee2mqtt/issues/9929, I need to change which entities are exposed to mqtt/Home assistant dynamically. I tried to be as less intrusive as possible, taking in account the current functionality of the rest of devices/integrations that already existed.

This PR consists of 3 main features:

- ~~A  `start` step (optional) which can be used by any device to run something every time Z2M is started. If it doesn't exists, is ignored.~~ Using onEvent instead.
- A attribute visible (true by default) which is currently used as mechanism to expose or not a entitity (https://github.com/Koenkk/zigbee-herdsman-converters/pull/3492)
- ~~Exposing the specific options from the user to the `configure` step (optional too).~~ Using events too

The goal is that a user could control with specific options the behaviour of the integration, which could change what it is exposing based on the user options.

Since the attributes are sent to MQTT with `publishDevices()`, the user had to reconfigure the zigbee device in order to apply the changes

Koenkk commented on lib/extension/bridge.ts:
```diff
@@ -598,6 +600,18 @@ export default class Bridge extends Extension {
         return Object.values(scenes);
     }
 
+    async startDevices(): Promise<void> {
```
Something like start devices is already available in the form of `onEvent` with `type` `start`, example: https://github.com/Koenkk/zigbee-herdsman-converters/blob/ee8c33be9656c4982889b49319b7236deef6bd92/devices/tuya.js#L1104

vk496 replied:
Oh, thnks! That simplify this PR. I will update it

vk496 replied:
Hi,

I checked it again. The reason I did it like that is because I need to execute the `start` step BEFORE the publishDevices(), since I toggle there which entities are exposed.

Koenkk replied:
I believe that can be fixed by moving https://github.com/Koenkk/zigbee2mqtt/blob/42569d6c39295b6bcbde788a2db94e0262b434e8/lib/controller.ts#L76 above `ExtensionBridge`

vk496 replied:
Hi,

That was what I tried yesterday. But since the initialization is donde concurrently, it may came first or may not.

And I was not sure proposing to do it  sequentially...

Koenkk replied:
Moving https://github.com/Koenkk/zigbee2mqtt/blob/42569d6c39295b6bcbde788a2db94e0262b434e8/lib/extension/onEvent.ts#L10 to the constructor of `onEvent` should fix it.

vk496 replied:
Finally I called it `devicesStarted` and waited all promises before trigger it. I couldn't test it in real environment but it should work even if some `onEvent` fails for some reason...

Koenkk commented:
Note to self; after this continue with:
- https://github.com/Koenkk/zigbee-herdsman-converters/pull/3493
- https://github.com/Koenkk/zigbee2mqtt/pull/10133
- https://github.com/Koenkk/zigbee-herdsman-converters/pull/3492

vk496 committed:
Dynamically expose attributes

vk496 commented:
Hello @Koenkk 

I modified the code in order to trigger a event after the `start` callback. I also adapted the code to pass all unit tests. I just missed the `Should publish devices on startup` because I was not sure how to adapt the unit and use 2 extensions (Bridge and OnEvent) properly.

Hope is good enough to merge it to dev. There are some people waiting these PRs in order to start testing the zlinky integration

Koenkk commented on lib/types/types.d.ts:
```diff
@@ -94,15 +94,16 @@ declare global {
             endpoint?: string, values?: string[], value_off?: string, value_on?: string,
             access: number, property: string, unit?: string,
-            value_min?: number, value_max?: number}
+            value_min?: number, value_max?: number, visible: boolean}
```
IMO it would be better to keep the `visible` thing internal. Meaning it is not exposed to MQTT (so `Should publish devices on startup` and `Should publish event and devices when device interview successful` testcases should remain unchanged) and the filtering is done on the z2m side.

vk496 replied:
But if I don't define the attribute here, I can't use it/reference it in the rest of the project.

I'm wrong?

Koenkk replied:
I shouldn't have put the comment on this line indeed, sorry for that. The removing of the `visible` attribute can be done in `bridge.ts`.

Also what do you think about naming it `enabled` instead of `visible`?

Koenkk commented on lib/extension/configure.ts:
```diff
@@ -112,7 +112,8 @@ export default class Configure extends Extension {
         try {
-            await device.definition.configure(device.zh, this.zigbee.firstCoordinatorEndpoint(), logger);
+            await device.definition.configure(device.zh,
+                this.zigbee.firstCoordinatorEndpoint(), logger, device.settings);
```
I see options are used here https://github.com/Koenkk/zigbee-herdsman-converters/pull/3493/files#diff-b8ac9ade6e7ce3d01aa9354bbe562e0277f0d1d9826fcc041ea9cd55a63313cbR667 for the dynamic expose stuff, however it would be nicer to add a new onEvent with even "deviceJoined" (so this can be removed again)

vk496 replied:
ACK

Koenkk commented on lib/extension/onEvent.ts:
```diff
@@ -1,15 +1,24 @@
         }
 
+        Promise.all(myPromises)
+            .then(() => this.eventBus.emitDevicesStarted())
```
You can use the already existing `emitDevicesChanged` event for this.

vk496 replied:
Do you think is ok to use it? I saw is already used for other events too

Koenkk commented on lib/extension/bridge.ts:
```diff
@@ -684,11 +687,20 @@ export default class Bridge extends Extension {
             vendor: device.definition.vendor,
             description: device.definition.description,
-            exposes: device.definition.exposes,
+            exposes: filteredExposes,
```
This can be simplified to `exposes: device.definition.exposes.filter((e) => e.visible),` (so the `treatExposes` can be removed)

vk496 replied:
Yes, that was my initial proposal. But after reviewing the unit tests, `exposes` could be null in some cases.

If you know a cool ES6 way to do it, I will be glad to learn it

Koenkk replied:
For which cases exposes is null? That shouldn't happen.

vk496 replied:
I don't remember it right now. But it was in a unit test (I can check it again this afternoon)

Koenkk commented on lib/controller.ts:
```diff
@@ -263,6 +263,31 @@ class Controller {
+
+        // filter not visible attributes
+        if (entity.isDevice() && entity.definition && entity.definition.exposes) {
+            // We could have 2 different entities with same attribute and different visibility.
```
Could you give an example for this?

vk496 replied:
The zlinky device emit `BASE` (https://github.com/vk496/zigbee-herdsman-converters/blob/zlinky-integration/devices/lixee.js#L513) when is operating in "Historique/Legacy" mode and `EAST` (https://github.com/vk496/zigbee-herdsman-converters/blob/zlinky-integration/devices/lixee.js#L553) when operating in "Standard" mode.

Both are using the same cluster attribute (`currentSummDelivered`) and are mutually exclusive. So I can't just make visible/hide the cluster attribute and must filter by entity instead

Koenkk replied:
But at this point we have nothing to do anymore with the cluster attributes right? Since everything has already been converted in the fromZigbee converters. So lets say `BASE` is visible and `EAST` isn't, the converter returns `{"BASE": 123, "EAST": 123}`, since `EAST` isn't visible we just have to remove that here.

vk496 committed:
Rollback options in configure step

Use event instead, as proposed in https://github.com/Koenkk/zigbee2mqtt/pull/10132/files#r771801648

vk496 committed:
Simplify code

vk496 committed:
Switch to already existing event

vk496 committed:
Fix all unit tests

vk496 committed:
Remove redundant code

vk496 committed:
Perfect rollback

vk496 commented:
/ping @Koenkk 

Ready with all unit test passing

Koenkk commented:
Sorry for noticing this just now, but what happens when you have 2 devices with different configuration (exposes), if I understand correctly this won't work now. (if this is indeed the case, I can provide a clean solution for this)

vk496 commented:
Sorry, I'm not fully understand you. What do you mean?

When a exposes change the visibility of some entity, it always refer to himself (or at least, what I did in the PR for zlinky).

Btw, 2 things to note:

- This PR will fail if visibility PR is not merged first in Z2M-converters.
- I'm not able to create a unit test for the missing coverage of onEvent.ts (lines 17-20). Sorry, already tried but don't really know how to do it.

PS: I will try to fix the TS build this night

vk496 committed:
Fix typescript types

Koenkk commented:
The idea is to either have exposes as an array OR as a function. From this function you can return whatever exposes you want (based on the `settings`) so the interface will be this:

```js
exposes: DefinitionExpose[] | ((device: zh.Device, settings: KeyValue) => DefinitionExpose[])
```

I've commited this first part to this PR, but still need to do some refactoring in `homeassistant.ts` before I can continue.

Koenkk committed:
Updates

Koenkk committed:
Updates

Koenkk committed:
Merge branch 'dev' into expose-dynamic

Koenkk committed:
Updates

Koenkk committed:
Merge branch 'dev' into expose-dynamic

Koenkk committed:
updates

Koenkk commented:
@vk496 ready now, could you test it? If it works this can be merged. Note that no changes to zigbee-herdsman-converters are needed.

So to summarise, exposes can be a method now, example:
```js
...
vendor: 'myvendor',
description: 'my description'
exposes: (device, options) => {
    // Return list of exposes here.
    // NOTE: it should also return something when device and options are null (required for generating the docs)
    return [e.power_outage_memory()];
},
...
```

Koenkk committed:
Improve

vk496 commented:
@Koenkk Thanks for your help. In my side all OK except publishDevice() issue

~~(Reminder: in order to configure dynamically the exposition of the entity, the `start` event must be executed **before** publishDevices() or mqtt will expose both)~~ EDIT: Since the expose call pass the device and options, that is no longer a requirement. Green light :+1:

vk496 commented:
`configure:` must expose `options` too, since we must know to which entities we had to subscribe. It may work with `devicesChanged`, but I'm not sure because a successful reporting configuration already emit this event, wo we may enter in a infinite loop. I think is easier to pass `options` as optional parameter to `configure`

Koenkk commented:
@vk496 the problem with passing options to `configure()` is that `configure()` is not called again once the options change. It would be better to pass the options to the `onEvent` so that you can correctly configure the device with the onEvent `start`

vk496 commented:
Ok, you're right, sorry. So I will remove it.

I will come back if I have any troubles with zlinky integration, but at least with this I can start telling people to use it

vk496 commented:
Ready to merge. Thank you for your help :dancers:

Koenkk committed:
Merge branch 'dev' into expose-dynamic

Koenkk committed:
Merge branch 'dev' into expose-dynamic
